fletcher/algorithms/string.py

```diff
@@ -333,54 +333,93 @@ def _slice_handle_chunk(pa_arr, start, end, step):
+        # if step == 0:
 
-        char_idx = 0
-        byte_idx = 0
+        str_len_bytes = offsets[i + 1] - offsets[i]
 
+        # To do: add more empty cases
         if start > str_len_bytes:
             builder.append_empty()
             continue
 
-        while char_idx < start and byte_idx < str_len_bytes:
-            char_idx += 1
-            byte_idx += get_utf8_size(data[offsets[i] + byte_idx])
+        # Positive start, end, and step
+        if start >= 0 and (end is None or end >= 0) and step > 0:
+            char_idx = 0
+            byte_idx = 0
+
+            while char_idx < start and byte_idx < str_len_bytes:
+                char_idx += 1
+                byte_idx += get_utf8_size(data[offsets[i] + byte_idx])
 
-        # positive start, end; step > 1
-        if step > 1:
-            to_skip = 0
-            include_bytes: List[bytes] = []
+            # Step = 1
+            if step == 1:
+                start_byte = offsets[i] + byte_idx
 
-            while char_idx < end and byte_idx < str_len_bytes:
-                char_size = get_utf8_size(data[offsets[i] + byte_idx])
+                while (end is None or char_idx < end) and byte_idx < str_len_bytes:
+                    char_idx += 1
+                    byte_idx += get_utf8_size(data[offsets[i] + byte_idx])
 
-                if not to_skip:
-                    include_bytes.extend(
-                        data[offsets[i] + byte_idx : offsets[i] + byte_idx + char_size]
-                    )
-                    to_skip = step
+                end_byte = offsets[i] + byte_idx
+                builder.append_value(data[start_byte:end_byte], end_byte - start_byte)
 
-                char_idx += 1
+            # Step > 1
+            else:
+                to_skip = 0
+                include_bytes: List[bytes] = []
+
+                while (end is None or char_idx < end) and byte_idx < str_len_bytes:
+                    char_size = get_utf8_size(data[offsets[i] + byte_idx])
+
+                    if not to_skip:
+                        include_bytes.extend(
+                            data[
+                                offsets[i]
+                                + byte_idx : offsets[i]
+                                + byte_idx
+                                + char_size
+                            ]
+                        )
+                        to_skip = step
+
+                    char_idx += 1
+                    byte_idx += char_size
+                    to_skip -= 1
+
+                builder.append_value(include_bytes, len(include_bytes))
+
+        # Other: negative start, end, or step
+        else:
+            char_bytes: List[int] = []
+            byte_idx = 0
+
+            while byte_idx < str_len_bytes:
+                char_size = get_utf8_size(data[offsets[i] + byte_idx])
+                char_bytes.append(
+                    data[offsets[i] + byte_idx : offsets[i] + byte_idx + char_size]
+                )
```

Comment:
I think this is wrong here. You don't have a list of int but a list of whatever you get when you slice the data object.